src/node/rpc/nodefrontend.h

```diff
@@ -1,95 +1,206 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the Apache 2.0 License.
-#include "../../crypto/hash.h"
-#include "../../enclave/oe_shim.h"
-#include "../entities.h"
-#include "../quoteverification.h"
+#include "crypto/hash.h"
 #include "frontend.h"
+#include "node/entities.h"
+#include "node/networkstate.h"
+#include "node/quoteverification.h"
 
 namespace ccf
 {
-  class NodesCallRpcFrontend : public RpcFrontend
+  class NodeCallRpcFrontend : public RpcFrontend
   {
+  private:
+    NetworkState& network;
+    AbstractNodeState& node;
+
+    std::optional<NodeId> check_node_exists(
+      Store::Tx& tx,
+      std::vector<uint8_t>& node_pem,
+      std::optional<NodeStatus> node_status = std::nullopt)
+    {
+      auto nodes_view = tx.get_view(network.nodes);
+
+      std::optional<NodeId> existing_node_id;
+      nodes_view->foreach([&existing_node_id, &node_pem, &node_status](
+                            const NodeId& nid, const NodeInfo& ni) {
+        if (
+          ni.cert == node_pem &&
+          (!node_status.has_value() || ni.status == node_status.value()))
+        {
+          existing_node_id = nid;
+          return false;
+        }
+        return true;
+      });
+
+      return existing_node_id;
+    }
+
+    std::optional<NodeId> check_conflicting_node_network(
+      Store::Tx& tx, const NodeInfoNetwork& node_info_network)
+    {
+      auto nodes_view = tx.get_view(network.nodes);
+
+      std::optional<NodeId> duplicate_node_id;
+      nodes_view->foreach([&node_info_network, &duplicate_node_id](
+                            const NodeId& nid, const NodeInfo& ni) {
+        if (
+          node_info_network.nodeport == ni.nodeport &&
+          node_info_network.host == ni.host && ni.status != NodeStatus::RETIRED)
+        {
+          duplicate_node_id = nid;
+          return false;
+        }
+        return true;
+      });
+
+      return duplicate_node_id;
+    }
+
+    auto add_node(
+      Store::Tx& tx,
+      std::vector<uint8_t>& caller_pem_raw,
+      const JoinNetworkNodeToNode::In& in,
+      NodeStatus node_status)
+    {
+      auto nodes_view = tx.get_view(network.nodes);
+
+      auto conflicting_node_id =
+        check_conflicting_node_network(tx, in.node_info_network);
+      if (conflicting_node_id.has_value())
+      {
+        return jsonrpc::error(
+          jsonrpc::StandardErrorCodes::INVALID_PARAMS,
+          fmt::format(
+            "A node with the same node host {} and port {} already "
+            "exists "
+            "(node id: {})",
```

Comment:
(V Minor: This doesn't need a separate line. `clang-format` splits these lines but won't reform them later)